### PR TITLE
Stripe assessment tool [RAM-247]

### PR DESCRIPTION
**Pull request by DigitalSuey**

### CODE REVIEW

**Comment by DigitalSuey** [comment]
To do:

- [ ] Calculate and present results in table
- [ ] The purple answer selection is only working on the first answers section
- [ ] Update Ben's script so that it exports json data in the correct format as assessment2.json
- [ ] Update the spread sheet to contain a link to the results the user sees

**Comment by DigitalSuey** [comment]
⚠️ **NOTE** ⚠️ 

Before merging this, don't forget to change the email links from staging to production.

File: _src/controllers/assessment-controller.js [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+    xhr.onreadystatechange = function() {
+      if (xhr.readyState === 4 && xhr.status === 200) {
+        self.formTarget.reset();
+        self.formTarget.classList.toggle("d-none");
+        self.responseTarget.classList.toggle("d-none");
+        self.responseTarget.classList.toggle("d-flex");
+      }
+    };
+
+    var encoded = Object.keys(formData.data).map(function(k) {
```
Comment (jsntv200) [inline _src/controllers/assessment-controller.js:242]: Fairly sure [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) or something similar has methods for this.

File: _src/controllers/assessment-controller.js [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    this.navigateTo(event.params.href);
+  }
+
+  setActiveAnswer() {
+    const answers = this.getAnswers();
+    const section = location.pathname.split('/').reverse()[0];
+  
+    if (Object.entries(answers).length === 0) return;
+  
+    for (const i in this.answerTargets) {
```
Comment (jsntv200) [inline _src/controllers/assessment-controller.js:273]: I think for of will work here:

```javascript
for (const target of this.answerTargets) {
  ...
}
```

File: _src/controllers/assessment-controller.js [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+          this.sectionTargets[i].removeAttribute("disabled");
+          this.sectionTargets[i].classList.remove("opacity-25");
+
+          this.sectionSliderTarget.scrollLeft = this.sectionTargets[i].offsetLeft - this.sectionSliderTarget.offsetLeft;
+        }
+      }
+    }
+  }
+
+  setAnswer(data) {
```
Comment (jsntv200) [inline _src/controllers/assessment-controller.js:329]: Has to be a better way to do this.

File: _src/controllers/assessment-controller.js [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    } else {
+      answers[answer[0]] = [[parseInt(answer[2]), parseInt(answer[3])]];
+    }
+
+    // Saved data format: {"sectionName": [["answerNumber", "answerWeightNumber"], ["answerNumber", "answerWeightNumber"]]}
+    // The answer positions in the setionName array correspond to the question number
+    localStorage.setItem("answers", JSON.stringify(answers));
+  }
+
+  setResult(sectionIndex, resultIndex) {
```
Comment (jsntv200) [inline _src/controllers/assessment-controller.js:349]: Has to be a better way to do this.

**Review by jsntv200 — COMMENTED** [review]
Added a bunch of comments here just for consideration to clean up the code and best practices. You don't have to implement as I assume its all working, but remember to ask yourself will future me understand this or someone else.

File: _src/controllers/assessment-controller.js [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    "submitButton"
+  ];
+
+  static values = {
+    basePath: "/online-payments/payment-maturity-assessment/",
+    colorHashes: {type: Array, default: ["dc697a", "fdc95b", "c2d7b1", "92defb", "9069f7"]},
+    sections: {type: Array, default: ["current", "your-team", "owner", "technology"]}
+  }
+
+  connect() {
```
[the remark this answers is not in the document]
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js:31]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    } else {
+      return Math.abs( x ) + Math.abs( y );
+    }
+  }
+
+  download() {
+    window.print();
+  }
+
+  getAnswers() {
```
[the remark this answers is not in the document]
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+    if (Object.entries(answers).length > 0) {
+      this.resumeTarget.classList.toggle("d-none");
+    }
+  }
+  
+  clearAnswers() {
+    localStorage.removeItem("answers");
+  }
+
+  clearErrors() {
```
[the remark this answers is not in the document]
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js:139]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+      }
+    }).filter(function(item, pos, self) {
+      return self.indexOf(item) == pos && item;
+    });
+
+    fields.forEach(function(name) {
+      const element = elements[name];
+      formData[name] = element.value;
+
+      if (elements[name].name === "email") {
```
[the remark this answers is not in the document]
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+        email = formData[name];
+      }
+
+      if (element.length) {
+        var data = [];
+        
+        for (var i = 0; i < element.length; i++) {
```
[the remark this answers is not in the document]
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+        formData[name] = data.join(', ');
+      }
+    });
+
+    // add form-specific values into the data
+    formData.formDataNameOrder = JSON.stringify(fields);
+    formData.formGoogleSheetName = this.formTarget.dataset.sheet || "responses";
+    formData.formGoogleSendEmail = email;
+
+    return {data: formData, honeypot: honeypot};
```
[the remark this answers is not in the document]
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+  }
+
+  getParams() {
+    return new Map(location.search.slice(1).split('&').map(kv => kv.split('=')));
```
[the remark this answers is not in the document]
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+
+    var xhr = new XMLHttpRequest();
+
+    xhr.open('POST', url);
+    xhr.setRequestHeader("Content-Type", "application/x-www-form-urlencoded");
+    xhr.onreadystatechange = function() {
+      if (xhr.readyState === 4 && xhr.status === 200) {
+        self.formTarget.reset();
+        self.formTarget.classList.toggle("d-none");
+        self.responseTarget.classList.toggle("d-none");
```
[the remark this answers is not in the document]
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js (hunk 1 above)
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js:251]: Updated in [9a7fb07](https://github.com/red-ant/redant-home/pull/302/commits/9a7fb07c3aabfc8fe5069a8e8a405dd73f122f29).

File: _src/controllers/assessment-controller.js (hunk 2 above)
Comment (DigitalSuey) [reply _src/controllers/assessment-controller.js:279]: I still need the index further down the loop.

**Pull request merged**